setuptools/command/egg_info.py

```diff
@@ -293,20 +295,35 @@ def delete_file(self, filename):
             os.unlink(filename)
 
     def run(self):
-        self.mkpath(self.egg_info)
-        try:
-            os.utime(self.egg_info, None)
-        except OSError as e:
-            msg = f"Cannot update time stamp of directory '{self.egg_info}'"
-            raise distutils.errors.DistutilsFileError(msg) from e
-        for ep in metadata.entry_points(group='egg_info.writers'):
-            writer = ep.load()
-            writer(self, ep.name, os.path.join(self.egg_info, ep.name))
-
-        # Get rid of native_libs.txt if it was put there by older bdist_egg
-        nl = os.path.join(self.egg_info, "native_libs.txt")
-        if os.path.exists(nl):
-            self.delete_file(nl)
+        # Avoid adding an empty .egg-info to sys.path while using importlib.metadata
+        # See pypa/pyproject-hooks#206
+        self.mkpath(self.egg_base)  # avoid file system errors with tmpdir / os.replace
+        with TemporaryDirectory(prefix=".tmp-", dir=self.egg_base) as tmp:
+            staging = os.path.join(tmp, "egg_info")
+            if os.path.isdir(self.egg_info):
+                shutil.copytree(self.egg_info, staging)  # account pre-existing files
+            else:
+                os.mkdir(staging)
+
+            for ep in metadata.entry_points(group='egg_info.writers'):
+                writer = ep.load()
+                writer(self, ep.name, os.path.join(staging, ep.name))
+
+            # Get rid of native_libs.txt if it was put there by older bdist_egg
+            nl = os.path.join(staging, "native_libs.txt")
+            if os.path.exists(nl):  # pragma: no cover
+                self.delete_file(nl)
+
+            # Remove old directory and create the new one
+            try:
+                # Unfortunately os.replace does not work for existing destination dirs,
+                # so we cannot have a single atomic operation
+                _shutil.rmdir(self.egg_info)
+                os.replace(staging, self.egg_info)
+                log.info(f"renaming {staging!r} to {self.egg_info!r}")
+            except OSError as e:
+                msg = f"Cannot create directory '{self.egg_info}' ({e})"
+                raise distutils.errors.DistutilsFileError(msg) from e
```

Comment:
@jaraco in this rewrite the main idea is to use a temporary directory to avoid adding an empty `.egg-info` to the path right before iterating over `importlib_metadata.entry_points`.

May be a bit convoluted, but this is what I could come up to make the process a bit more atomic.

This is related to the discussion in https://github.com/pypa/pyproject-hooks/issues/206.